lib/bats-core/tracing.bash

```diff
@@ -117,15 +127,61 @@ bats_trim_filename() {
 	printf -v "$2" '%s' "${1#$BATS_CWD/}"
 }
 
+# bats_debug_trap tracks the last line of code executed within a test. This is
+# necessary because $BASH_LINENO is often incorrect inside of ERR and EXIT
+# trap handlers.
+#
+# Below are tables describing different command failure scenarios and the
+# reliability of $BASH_LINENO within different the executed DEBUG, ERR, and EXIT
+# trap handlers. Naturally, the behaviors change between versions of Bash.
+#
+# Table rows should be read left to right. For example, on bash version
+# 4.0.44(2)-release, if a test executes `false` (or any other failing external
+# command), bash will do the following in order:
+# 1. Call the DEBUG trap handler (bats_debug_trap) with $BASH_LINENO referring
+#    to the source line containing the `false` command, then
+# 2. Call the DEBUG trap handler again, but with an incorrect $BASH_LINENO, then
```

Comment:
I don't know that it's fair to call these "incorrect"; the DEBUG trap is firing prior to three different commands; the user's command, the ERR trap command, and the EXIT trap command. Are the line numbers actually incorrect, or just not what we care about?

Reply:
> I don't know that it's fair to call these "incorrect"

The behavior changed between Bash 4.0.44 and Bash 4.4.20, so my assumption is that Bash fixed a bug (or unintentionally changed behavior? unsure). I'm calling the Bash 4.0.44 behavior incorrect, compared to Bash 4.4.20's behavior.